Bug 1871108: usability bug: the ui selections are unclear

glekner opened the pull request:
As of now, I only updated the Provision Source dropdown so we can examine the differences.
Until now, we used a component which renders the browser's native select component

Example
<img width="710" alt="Screen Shot 2020-08-25 at 17 18 10" src="https://user-images.githubusercontent.com/24938324/91186118-4f4d5480-e6f7-11ea-81c2-0bd86986dd78.png">

For us to have a more robust UI with descriptions and other props, we need to convert to the new PF `Select` component, its a change that could break tests, and it could take time to change all the selects in the wizard.

1 Key difference between the components is the new Select rendering the dropdown to the bottom or top is not automatic like the native select. causing it to get hidden by the wizard footer: (changing z-index didn't fix the issue)
<img width="720" alt="Screen Shot 2020-08-25 at 17 15 47" src="https://user-images.githubusercontent.com/24938324/91186048-3b095780-e6f7-11ea-8fe3-b0a083365ba9.png">

The solution is to change the direction manually for some of the selects.
<img width="683" alt="Screen Shot 2020-08-25 at 17 12 29" src="https://user-images.githubusercontent.com/24938324/91186089-46f51980-e6f7-11ea-9359-801cc2ffefc4.png">

glekner commented:
@suomiy @matthewcarleton thoughts?

matthewcarleton commented:
I guess an alternative to this would be to scroll the user down so the whole dropdown is visible.

glekner commented:
* OS and Flavor components are now separated.
* Source is now between them

Some of the dropdown's content is still hidden by the footer but at least the user will know to scroll.

@suomiy @yaacov @irosenzw can you review

matthewcarleton commented:
@glekner Is there a reason why the selection titles don't align with the design?
![image](https://user-images.githubusercontent.com/287880/91324924-fd5d0b00-e798-11ea-9a17-123a34dd5a6c.png)

yaacov commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/vm-settings-tab.tsx:
```diff
@@ -128,6 +114,34 @@ export const VMSettingsTabComponent: React.FC<VMSettingsTabComponentProps> = ({
         onProvisionSourceStorageChange={updateStorage}
         provisionSourceStorage={provisionSourceStorage}
       />
+      <Flavor
```
I like that we have flavor+mem+cpu together :+1:

suomiy commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/flavor.tsx:
```diff
@@ -0,0 +1,113 @@
+  userTemplates: any;
+  commonTemplates: any;
+  flavorField: any;
+  operatinSystemField: any;
```
you can pass just the value, not the whole field

suomiy commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/os.tsx:
```diff
@@ -42,7 +40,7 @@ import {
+export const OS: React.FC<OSProps> = React.memo(
   ({
     userTemplates,
     commonTemplates,
```
same here with the flavorField

suomiy commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/flavor.tsx:
```diff
@@ -0,0 +1,113 @@
+  commonTemplates: any;
+  flavorField: any;
+  operatinSystemField: any;
+  cloneBaseDiskImageField: any;
```
no need for these

glekner commented:
* Source dropdown now overlaps the wizard footer
* removed unused props

can you approve @suomiy

jelkosz commented:
@glekner One small comment: please enrich the PXE part by a mention that you need a network attachment definition available for you in this namespace or in the default one. So the message might be like:
"Boot an operating system from a server on a network. This will show up as a NIC on the Networking step. In order to PXE boot, you need to have a Network Attachment Definition available for you either in this namespace, or in the default one."

suomiy commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/provision-source.tsx:
```diff
@@ -69,19 +75,33 @@ export const ProvisionSourceComponent: React.FC<ProvisionSourceComponentProps> =
-            <FormSelectPlaceholderOption
-              placeholder={getPlaceholder(VMSettingsField.PROVISION_SOURCE_TYPE)}
-              isDisabled={!!provisionSourceValue}
+        <Select
```
can you please create a new type for FormField and use it here? This will allow to pass correct values and metadata from redux to this component.

glekner commented:
@jelkosz updated PXE text
@suomiy added `PF_SELECT` field type. the only supported prop I can see is `isDisabled`

glekner commented:
/retest

suomiy commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/provision-source.tsx:
```diff
@@ -69,18 +75,34 @@ export const ProvisionSourceComponent: React.FC<ProvisionSourceComponentProps> =
-            />
+          <Select
+            isOpen={isSourceOpen}
+            selections={provisionSourceValue}
```
we can pass this one in the FormField as well

suomiy commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/provision-source.tsx:
```diff
@@ -69,18 +75,34 @@ export const ProvisionSourceComponent: React.FC<ProvisionSourceComponentProps> =
+          <Select
+            isOpen={isSourceOpen}
+            selections={provisionSourceValue}
+            onToggle={(isOpen) => setIsSourceOpen(isOpen)}
```
maybe it would be beneficial to create a single select component that has all the defaults included and stores its own state.
This will come in handy once we include multiple of these in.

jelkosz commented on frontend/packages/kubevirt-plugin/src/utils/strings.ts:
```diff
@@ -18,6 +18,16 @@ export const ADD_NETWORK_INTERFACE = 'Add Network Interface';
+export const PROVISION_SOURCE_CONTAINER_HELPTEXT =
+  'Link to a bootable operating system container located in a registry accessible from the cluster. Example: kubevirt/cirrosregistry-disk-demo. This action will not create a persistent volume claim (PVC). This will show up as a disk in the Storage step';
+export const PROVISION_SOURCE_URL_HELPTEXT =
+  'Link to an operating system image via URL (HTTP or S3 endpoint). A new persistent volume claim (PVC) will be created to store this image. This will show up as a disk in the Storage step';
```
Maybe would be good to add an example as well, something like:

"For example, for CentOS you can visit http://isoredirect.centos.org/centos/8/isos/x86_64/ and pick a URL for an ISO which will look like: http://mirror.hosting90.cz/centos/8.2.2004/isos/x86_64/CentOS-8.2.2004-x86_64-boot.iso"

@matthewcarleton WDYT? Would something like this be helpful?

jelkosz replied:
OK, that might be a bit too much and something which will be changing so we should not hardcode it here. But we could have something which will help. Something like:
"Link to an operating system image via URL (HTTP or S3 endpoint). An example might be a download link to an operating system iso you can copy from the downloads page of your favourite distribution. A new persistent volume claim (PVC) will be created to store this image. This will show up as a disk in the Storage step.".

@matthewcarleton WDYT? Would this be understandable?

matthewcarleton replied:
Maybe:

Link to an operating system source via URL (HTTP or S3). This will show up as a disk in the storage step (backed by a persistent volume claim). **Example: A operating system ISO download link you can copy from your favorite distribution**.

For brevity.

suomiy replied:
can we also specify the formats?

jelkosz replied:
I would explicitely add the download page there: Link to an operating system source via URL (HTTP or S3). This will show up as a disk in the storage step (backed by a persistent volume claim). Example: An operating system ISO download link you can copy from the download page of your favorite distribution.

glekner commented:
@suomiy 
* added labels to `ProvisionSource`
* created `FormPFSelect` component

yaacov commented:
/retest

matthewcarleton commented:
> @glekner One small comment: please enrich the PXE part by a mention that you need a network attachment definition available for you in this namespace or in the default one. So the message might be like:
> "Boot an operating system from a server on a network. This will show up as a NIC on the Networking step. In order to PXE boot, you need to have a Network Attachment Definition available for you either in this namespace, or in the default one."

I put this together which could solve this
<img width="571" alt="Screen Shot 2020-09-02 at 9 44 25 AM" src="https://user-images.githubusercontent.com/287880/91982613-e53d3c80-ed00-11ea-8521-f4fdcd702ad1.png">

glekner commented:
* Added `FormPFSelect` component
* Created `SelectDropdownObjectEnum` class for abstraction and reusability.
* Changed OS, Flavor, Workload, Disk Source, NIC Model, NIC Type to use the new component.
* Flavor, Workload don't have descriptions as of now.

<img width="680" alt="Screen Shot 2020-09-02 at 18 36 10" src="https://user-images.githubusercontent.com/24938324/92004961-8e108980-ed4b-11ea-8e76-12493128533d.png">

<img width="639" alt="Screen Shot 2020-09-02 at 18 35 46" src="https://user-images.githubusercontent.com/24938324/92004874-70432480-ed4b-11ea-9200-b4418f5fef66.png">

<img width="634" alt="Screen Shot 2020-09-02 at 18 38 21" src="https://user-images.githubusercontent.com/24938324/92004938-851fb800-ed4b-11ea-8cfa-73ff72553c09.png">

<img width="631" alt="Screen Shot 2020-09-02 at 18 38 29" src="https://user-images.githubusercontent.com/24938324/92004943-86e97b80-ed4b-11ea-9676-a9a887eb5436.png">

suomiy commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/provision-source.tsx:
```diff
@@ -70,17 +70,32 @@ export const ProvisionSourceComponent: React.FC<ProvisionSourceComponentProps> =
+            {(sources || [])
+              .sort(
+                (a, b) =>
+                  ProvisionSource.fromString(a).getOrder() -
```
you can do the map first

glekner replied:
don't think so, the map returns a different array with `[<SelectOption>]`

suomiy commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/os.tsx:
```diff
@@ -193,22 +178,25 @@ export const OSFlavor: React.FC<OSFlavorProps> = React.memo(
               {operatingSystemBaseImages.map(({ id, name, message }) => (
-                <FormSelectOption key={id} value={id} label={`${name || id} ${message}`} />
+                <SelectOption key={id} value={id}>
+                  {`${name || id} ${message}`}
```
can render an extra space

suomiy replied:
because the message can be empty

suomiy commented on frontend/packages/kubevirt-plugin/src/constants/select-dropdown-object-enum.ts:
```diff
@@ -0,0 +1,28 @@
+    this.order = order;
+  }
+
+  getLabel = () => this.label;
```
can we use toString method here instead to be consistent. We use this approach in other places as well.

suomiy commented on frontend/packages/kubevirt-plugin/src/utils/strings.ts:
```diff
@@ -18,6 +18,38 @@ export const ADD_NETWORK_INTERFACE = 'Add Network Interface';
+export const PROVISION_SOURCE_PXE_DESC =
+  'Boot an operating system from a server on a network. This will show up as a NIC on the Networking step. Requires a PXE bootable network attachment definition';
+export const PROVISION_SOURCE_CONTAINER_DESC =
+  'Link to a bootable operating system container located in a registry accessible from the cluster. Example: kubevirt/cirros-registry-disk-demo. This action will not create a persistent volume claim (PVC). This will show up as a disk in the Storage step';
```
Do we have to specify this will not create a PVC? Kubernetes use should probably know what container is.

suomiy commented on frontend/packages/kubevirt-plugin/src/utils/strings.ts:
```diff
@@ -18,6 +18,38 @@ export const ADD_NETWORK_INTERFACE = 'Add Network Interface';
+
+// storage ui sources descriptions
+export const UI_SOURCE_BLANK_DESC = 'Create an empty disk (PVC)';
+export const UI_SOURCE_URL_DESC =
```
+formats?

suomiy commented on frontend/packages/kubevirt-plugin/src/utils/strings.ts:
```diff
@@ -18,6 +18,38 @@ export const ADD_NETWORK_INTERFACE = 'Add Network Interface';
+export const PROVISION_SOURCE_URL_DESC =
+  'Link to an operating system image via URL (HTTP or S3 endpoint). A new persistent volume claim (PVC) will be created to store this image. This will show up as a disk in the Storage step';
+export const PROVISION_SOURCE_DISK_DESC =
+  'Use a persistent volume claim (PVC) already available in the cluster. This will show up as a disk in the Storage step';
```
can we reuse the message already defined in UI_SOURCE_ATTACH_DISK_DESC and append the further description? Plust other types as well.

suomiy commented on frontend/packages/kubevirt-plugin/src/utils/strings.ts:
```diff
@@ -18,6 +18,38 @@ export const ADD_NETWORK_INTERFACE = 'Add Network Interface';
+
+// nics descriptions
+export const NIC_MODEL_VIRTIO_DESC =
+  'Optimized for best performance. Windows VMs requires virtio drivers to function.';
```
```suggestion
  'Optimized for best performance. Windows VMs require virtio drivers to function.';
```

suomiy commented on frontend/packages/kubevirt-plugin/src/utils/strings.ts:
```diff
@@ -18,6 +18,38 @@ export const ADD_NETWORK_INTERFACE = 'Add Network Interface';
+export const NIC_TYPE_BRIDGE_DESC =
+  'The VM will be bridged to the selected network, ideal for L2 devices';
+export const NIC_TYPE_SRIOV_DESC =
+  'Attach a virtual function network device to the VM for very high performance';
```
```suggestion
  'Attach a virtual function network device to the VM for high performance';
```
isn't this enough?

suomiy commented on frontend/packages/kubevirt-plugin/src/constants/vm/network/network-interface-model.ts:
```diff
@@ -1,15 +1,39 @@
+    order: 3,
+  });
+  static readonly NE2kPCI = new NetworkInterfaceModel('ne2kPCI', {
+    label: 'ne2kPCi',
```
small i? Let's rather default it to the value here

suomiy commented on frontend/packages/kubevirt-plugin/src/constants/vm/network/network-interface-model.ts:
```diff
@@ -1,15 +1,39 @@
+    order: 4,
+  });
+  static readonly PCNET = new NetworkInterfaceModel('pcnet', {
+    label: 'Pcnet',
```
```suggestion
    label: 'PCnet',
```

looks better?

gouyang commented on frontend/packages/kubevirt-plugin/src/utils/strings.ts:
```diff
@@ -18,6 +18,38 @@ export const ADD_NETWORK_INTERFACE = 'Add Network Interface';
+export const PROVISION_SOURCE_URL_DESC =
+  'Link to an operating system image via URL (HTTP or S3 endpoint). A new persistent volume claim (PVC) will be created to store this image. This will show up as a disk in the Storage step';
+export const PROVISION_SOURCE_DISK_DESC = `
+  ${UI_SOURCE_ATTACH_DISK_DESC}. This will show up as a disk in the Storage step`;
```
It's not showing up a disk in the Storage step.

gouyang commented on frontend/packages/kubevirt-plugin/src/utils/strings.ts:
```diff
@@ -18,6 +18,38 @@ export const ADD_NETWORK_INTERFACE = 'Add Network Interface';
+
+// provision sources descriptions
+export const PROVISION_SOURCE_PXE_DESC =
+  'Boot an operating system from a server on a network. This will show up as a NIC on the Networking step. Requires a PXE bootable network attachment definition';
```
The show up NIC is "Pod Networking" nic, it shows up for each provision source, maybe not worth mention it here.

jelkosz commented on frontend/packages/kubevirt-plugin/src/utils/strings.ts:
```diff
@@ -18,6 +18,16 @@ export const ADD_NETWORK_INTERFACE = 'Add Network Interface';
+export const PROVISION_SOURCE_PXE_HELPTEXT =
+  'Boot an operating system from a server on a network. This will show up as a NIC on the Networking step. In order to PXE boot, you need to have a Network Attachment Definition available for you either in this namespace, or in the default one';
+export const PROVISION_SOURCE_CONTAINER_HELPTEXT =
+  'Link to a bootable operating system container located in a registry accessible from the cluster. Example: kubevirt/cirrosregistry-disk-demo. This action will not create a persistent volume claim (PVC). This will show up as a disk in the Storage step';
```
So, we need to highlight the limitations of this and reference fedora instead of cirros:
Link to a bootable operating system container located in a registry accessible from the cluster. Example: kubevirt/fedora-cloud-container-disk-demo. This action will not create a persistent volume claim (PVC). This will show up as a disk in the Storage step. The container disk is meant to be used only for read-only filesystems such
as CD-ROMs or for small short-lived throw-away VMs.';

And then, once picked, under it it needs to show the example kubevirt/fedora-cloud-container-disk-demo so the user can actually copy paste it to the text box.

yaacov commented:
Gilad, hi, make sure that you didn't miss any `ProvisionSource[.].*[.]toString()` instance that should be `ProvisionSource[.].*[.]getValue()`.

It may be just some bad rebase, but while preparing the merged PR I found some.

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/provision-source.tsx:
```diff
@@ -26,9 +27,31 @@ const ProvisionSourceDiskHelpMsg: React.FC<ProvisionSourceDiskHelpMsgProps> = ({
+          <>
+            <Text> Enter URL here or edit the mounted disk in the {storageBtn} step</Text>
+            <Text>
+              For example, you can obtain the download link for {isUpstream ? 'Fedora' : 'RHEL'}{' '}
```
These examples should be moved below the relevant field. So in this case it would be under the URL input.

glekner replied:
the URL field is right below it. changing the help providers for each field is not in the scope.
we should merge the changes and if we still find it wrong, we should open a new bug.

matthewcarleton commented:
Can we change the "Source" label to "boot source"?

glekner commented:
- Rebased
- Changed title of provision source from 'Soruce' to 'Boot Source'

can we merge @yaacov @suomiy

yaacov commented:
/lgtm
/retest

yaacov commented:
/lgtm

yaacov commented:
/lgtm

yaacov commented:
/lgtm

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/tabs/vm-settings-tab/url-source.tsx:
```diff
@@ -34,6 +36,17 @@ export const URLSource: React.FC<URLSourceProps> = React.memo(
           />
         </FormField>
+        <div className="pf-c-form__helper-text" aria-live="polite">
+          For example, you can obtain the download link for {isUpstream ? 'Fedora' : 'RHEL'} cloud
```
can we change this to:
"Visit the Fedora cloud image[link] list and copy a url to paste into the url field here"

yaacov replied:
@glekner ^^

yaacov commented:
/lgtm
